### DISCUSSION TITLE
What's the point of using a loader to generate typings for CSS Modules?

### DISCUSSION THREAD

**Discussion by nyngwang** [webpack/webpack#18166] [Q&A] [answered]
I saw people writing loader to generate typings for CSS modules:

> When we add the webpack-typings-for-css loader, this will generate a TypeScript definition file ./src/button/style.scss.d.ts with the following content: [...]
> In your Typescript file you **will now be able to import this file and get type hints with the available classNames.**

From my understanding, a loader is run **after** we have finished writing our code and start the command, e.g. `webpack --watch --mode production`. So triggering compilation via webpack is the last step after each iteration during development: You write code with intellisense, save, and webpack compile.

With a loader to generate typings in the last step, we cannot get intellisense by those typings to write code conveniently. What's the point of using a loader to generate typings for CSS Modules then?

**Comment by alexander-akait** [comment]
To know the export of modules and all available names, this is not for IDE, this is to avoid mistakes

**Reply by nyngwang** [reply]
But when I'm writing the program, those `.d.ts` files haven't been produced by the loader. Then how could I avoid mistakes when writing code _before_ running a webpack-cli command to generating those `.d.ts` files?

**Reply by alexander-akait** [reply]
These files are virtual and generated by `webpack-typings-for-css loader`, you will never see them, it is just to know CSS classes

**Reply by nyngwang** [reply, marked as answer]
But one has to run `webpack --watch` in watch mode to keep generating new typings, correct? For example, in my workflow currently, ~~I will not run any webpack command before I finish a part of my program.~~

**update**: Now I understand that these loaders are meant for `--watch` mode.

**Reply by alexander-akait** [reply]
No, it is only for bundling process, not related to IDE at all
